### PR TITLE
Refactor device/track manager

### PR DESCRIPTION
**Pull request by czerwiukk**
## Description

Deduplication of device manager code, general improvements.
Groundwork for further improvements, especially `Client.ts` removal.

Feel invited to point out possible further improvements, it'd be best to write them down as separate tasks.

## Motivation and Context

This change strives to improve the codebase quality and maintainability.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
      not work as expected)

### CODE REVIEW

File: packages/react-client/src/Client.ts [hunk 1]
```diff
@@ -204,6 +217,7 @@ export interface ClientEvents<PeerMetadata, TrackMetadata> {
   managerStarted: (
     event: Parameters<DeviceManagerEvents["managerInitialized"]>[0] & {
       mediaDeviceType: MediaDeviceType;
+      trackType: TrackType;
     },
```
[the remark this answers is not in the document]
Comment (czerwiukk) [reply packages/react-client/src/Client.ts:221]: definitely beneficial, though I'd strive to remove this whole file in the future

File: packages/react-client/src/Client.ts [hunk 2]
```diff
@@ -559,55 +546,109 @@ export class Client<PeerMetadata, TrackMetadata> extends (EventEmitter as {
...
-      this.emit("deviceEnabled", { ...event, mediaDeviceType: "userMedia" }, this);
+      this.emit("deviceEnabled", { ...state, trackType: "audio", mediaDeviceType: "userMedia" }, this);
     });
 
-    this.deviceManager.on("managerInitialized", (event) => {
+    this.audioDeviceManager.on("managerInitialized", () => {
       this.stateToSnapshot();
 
-      this.emit("managerInitialized", { ...event, mediaDeviceType: "userMedia" }, this);
+      this.emit("managerInitialized", { mediaDeviceType: "userMedia" }, this);
```
Comment (kamil-stasiak) [inline packages/react-client/src/Client.ts:618]: Only the `userMedia` manager has an `init` method and can be initialized, so I think this` { mediaDeviceType: "userMedia" }` isn't very useful

File: packages/react-client/src/DeviceManager.ts [hunk 3]
```diff
@@ -1,563 +1,163 @@
...
-
-    if (audio.media?.deviceInfo) {
-      this.saveLastAudioDevice(audio.media?.deviceInfo);
-    }
+    const deviceInfo = this.deviceState.media?.deviceInfo;
+    if (deviceInfo) this.saveLastDevice(deviceInfo);
 
+    this.status = "initialized";
     this.setupOnEndedCallback();
+    this.emit("managerInitialized", this.deviceState);
```
Comment (kamil-stasiak) [inline packages/react-client/src/DeviceManager.ts:99]: Now this event is emitted twice (once for every device manager). Maybe it should be moved to `initializeDevices`

**Review by kamil-stasiak — CHANGES_REQUESTED** [review]

File: packages/react-client/src/mediaInitializer.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+type PreviousDevices = AudioVideo<MediaDeviceInfo | null>;
+
+const defaultErrors = { audio: null, video: null };
+
+const tryToGetVideoThenAudio = async (
+  constraints: MediaStreamConstraints,
+  deviceErrors: AudioVideo<DeviceError | null> = defaultErrors,
+  transformConstraint: (value?: boolean | MediaTrackConstraints) => boolean | MediaTrackConstraints | undefined = () =>
+    false,
+) => {
```
[the remark this answers is not in the document]
Comment (czerwiukk) [reply packages/react-client/src/mediaInitializer.ts]: resolved

**Review by kamil-stasiak — APPROVED** [review]

**Pull request merged**